File: copilot-to-api-main/copilot-client.js
#!/usr/bin/env node

const fs = require('fs');
const path = require('path');
const https = require('https');

class CopilotAPI {
    constructor(configPath = './config.json') {
        this.configPath = configPath;
        this.config = this.loadConfig();
        this.copilotToken = null;
        this.tokenExpiry = null;
    }

    loadConfig() {
        try {
            const configData = fs.readFileSync(this.configPath, 'utf8');
            return JSON.parse(configData);
        } catch (error) {
            console.error('Error loading config:', error.message);
            process.exit(1);
        }
    }

    async makeRequest(options, data = null) {
        return new Promise((resolve, reject) => {
            const req = https.request(options, (res) => {
                let body = '';
                res.on('data', chunk => body += chunk);
                res.on('end', () => {
                    try {
                        const response = JSON.parse(body);
                        resolve({ status: res.statusCode, data: response });
                    } catch (error) {
                        resolve({ status: res.statusCode, data: body });
                    }
                });
            });

            req.on('error', reject);
            
            if (data) {
                req.write(typeof data === 'string' ? data : JSON.stringify(data));
            }
            
            req.end();
        });
    }

    async refreshCopilotToken() {
        console.log('🔄 Refreshing Copilot token...');
        
        const options = {
            hostname: 'api.github.com',
            port: 443,
            path: '/copilot_internal/v2/token',
            method: 'GET',
            headers: {
                'authorization': `token ${this.config.access_token}`,
                'user-agent': 'GithubCopilot/1.155.0'
            }
        };

        try {
            const response = await this.makeRequest(options);
            
            if (response.status === 200) {
                this.copilotToken = response.data.token;
                // Use the expires_at timestamp from the API response (subtract 60 seconds for safety margin)
                this.tokenExpiry = (response.data.expires_at * 1000) - (60 * 1000);
                console.log('✅ Copilot token refreshed successfully');
                return true;
            } else {
                console.error('❌ Error refreshing token:', response.data);
                return false;
            }
        } catch (error) {
            console.error('❌ Network error refreshing token:', error.message);
            return false;
        }
    }

    async ensureValidToken() {
        if (!this.copilotToken || !this.tokenExpiry || Date.now() > this.tokenExpiry) {
            const success = await this.refreshCopilotToken();
            if (!success) {
                throw new Error('Failed to obtain valid Copilot token');
            }
        }
    }

    async getModels() {
        await this.ensureValidToken();
        
        const options = {
            hostname: 'api.githubcopilot.com',
            port: 443,
            path: '/models',
            method: 'GET',
            headers: {
                'authorization': `Bearer ${this.copilotToken}`,
                'Copilot-Integration-Id': 'vscode-chat'
            }
        };

        const response = await this.makeRequest(options);
        return response;
    }

    async chatCompletion(messages, options = {}) {
        await this.ensureValidToken();
        
        const requestData = {
            messages: messages,
            max_tokens: options.maxTokens || 1000,
            temperature: options.temperature || 0.3,
            stream: options.stream || false
        };

        const httpOptions = {
            hostname: 'api.githubcopilot.com',
            port: 443,
            path: '/chat/completions',
            method: 'POST',
            headers: {
                'authorization': `Bearer ${this.copilotToken}`,
                'Copilot-Integration-Id': 'vscode-chat',
                'content-type': 'application/json'
            }
        };

        const response = await this.makeRequest(httpOptions, requestData);
        return response;
    }
}

// Función para mostrar ayuda
function showHelp() {
    console.log(`
🤖 GitHub Copilot API Client (Node.js)

Usage:
  node copilot-client.js <command> [options]

Commands:
  models                     - Get available models
  chat "<message>"          - Send a chat message
  chat-file <file.txt>      - Send content from a file
  
Options:
  --max-tokens <number>     - Maximum tokens (default: 1000)
  --temperature <number>    - Temperature 0-1 (default: 0.3)
  --stream                  - Enable streaming

Examples:
  node copilot-client.js models
  node copilot-client.js chat "Write a Python function to sort a list"
  node copilot-client.js chat "Explain recursion" --max-tokens 500 --temperature 0.7
  node copilot-client.js chat-file prompt.txt

Note: Make sure to set your YOUR_ACCESS_TOKEN in config.json first!
    `);
}

// Función principal
async function main() {
    const args = process.argv.slice(2);
    
    if (args.length === 0 || args[0] === '--help' || args[0] === '-h') {
        showHelp();
        return;
    }

    const client = new CopilotAPI();
    const command = args[0];

    try {
        switch (command) {
            case 'models':
                console.log('📋 Getting available models...');
                const modelsResponse = await client.getModels();
                console.log('Models:', JSON.stringify(modelsResponse.data, null, 2));
                break;

            case 'chat':
                if (args.length < 2) {
                    console.error('❌ Please provide a message for chat');
                    return;
                }

                const message = args[1];
                const options = {};
                
                // Parse options
                for (let i = 2; i < args.length; i += 2) {
                    switch (args[i]) {
                        case '--max-tokens':
                            options.maxTokens = parseInt(args[i + 1]);
                            break;
                        case '--temperature':
                            options.temperature = parseFloat(args[i + 1]);
                            break;
                        case '--stream':
                            options.stream = true;
                            i--; // No value needed for this flag
                            break;
                    }
                }

                console.log('💬 Sending chat message...');
                const chatResponse = await client.chatCompletion([
                    { role: 'user', content: message }
                ], options);

                if (chatResponse.status === 200) {
                    console.log('\n📝 Response:');
                    console.log(chatResponse.data.choices[0].message.content);
                    console.log(`\n📊 Usage: ${chatResponse.data.usage.total_tokens} tokens`);
                } else {
                    console.error('❌ Error:', chatResponse.data);
                }
                break;

            case 'chat-file':
                if (args.length < 2) {
                    console.error('❌ Please provide a file path');
                    return;
                }

                const filePath = args[1];
                try {
                    const fileContent = fs.readFileSync(filePath, 'utf8');
                    console.log('📁 Sending file content...');
                    
                    const fileResponse = await client.chatCompletion([
                        { role: 'user', content: fileContent }
                    ]);

                    if (fileResponse.status === 200) {
                        console.log('\n📝 Response:');
                        console.log(fileResponse.data.choices[0].message.content);
                        console.log(`\n📊 Usage: ${fileResponse.data.usage.total_tokens} tokens`);
                    } else {
                        console.error('❌ Error:', fileResponse.data);
                    }
                } catch (error) {
                    console.error('❌ Error reading file:', error.message);
                }
                break;

            default:
                console.error('❌ Unknown command:', command);
                showHelp();
        }
    } catch (error) {
        console.error('❌ Error:', error.message);
    }
}

// Ejecutar si este archivo es llamado directamente
if (require.main === module) {
    main();
}

module.exports = CopilotAPI;

File: copilot-to-api-main/copilot_client.py
#!/usr/bin/env python3

import json
import time
import requests
import argparse
import os
from datetime import datetime, timedelta
from typing import Dict, List, Optional, Any

class CopilotAPI:
    def __init__(self, config_path: str = './config.json'):
        self.config_path = config_path
        self.config = self.load_config()
        self.copilot_token: Optional[str] = None
        self.token_expiry: Optional[datetime] = None

    def load_config(self) -> Dict[str, str]:
        """Load configuration from JSON file"""
        try:
            with open(self.config_path, 'r') as f:
                return json.load(f)
        except Exception as e:
            print(f'Error loading config: {e}')
            exit(1)

    def refresh_copilot_token(self) -> bool:
        """Refresh the Copilot token using the access token"""
        print('🔄 Refreshing Copilot token...')
        
        headers = {
            'authorization': f'token {self.config["access_token"]}',
            'user-agent': 'GithubCopilot/1.155.0'
        }

        try:
            response = requests.get(
                'https://api.github.com/copilot_internal/v2/token',
                headers=headers
            )
            
            if response.status_code == 200:
                data = response.json()
                self.copilot_token = data['token']
                # Use the expires_at timestamp from the API response (subtract 60 seconds for safety margin)
                self.token_expiry = datetime.fromtimestamp(data['expires_at'] - 60)
                print('✅ Copilot token refreshed successfully')
                return True
            else:
                print(f'❌ Error refreshing token: {response.text}')
                return False
        except Exception as e:
            print(f'❌ Network error refreshing token: {e}')
            return False

    def ensure_valid_token(self):
        """Ensure we have a valid Copilot token"""
        if not self.copilot_token or not self.token_expiry or datetime.now() > self.token_expiry:
            success = self.refresh_copilot_token()
            if not success:
                raise Exception('Failed to obtain valid Copilot token')

    def get_models(self) -> Dict[str, Any]:
        """Get available Copilot models"""
        self.ensure_valid_token()
        
        headers = {
            'authorization': f'Bearer {self.copilot_token}',
            'Copilot-Integration-Id': 'vscode-chat'
        }

        response = requests.get('https://api.githubcopilot.com/models', headers=headers)
        return {
            'status': response.status_code,
            'data': response.json() if response.status_code == 200 else response.text
        }

    def chat_completion(self, messages: List[Dict[str, str]], options: Optional[Dict[str, Any]] = None) -> Dict[str, Any]:
        """Send a chat completion request"""
        self.ensure_valid_token()
        
        if options is None:
            options = {}
        
        stream = options.get('stream', False)

        request_data = {
            'messages': messages,
            'max_tokens': options.get('max_tokens', 1000),
            'temperature': options.get('temperature', 0.3),
            'stream': stream
        }

        headers = {
            'authorization': f'Bearer {self.copilot_token}',
            'Copilot-Integration-Id': 'vscode-chat',
            'content-type': 'application/json'
        }

        response = requests.post(
            'https://api.githubcopilot.com/chat/completions',
            headers=headers,
            json=request_data,
            stream=stream
        )

        if not stream:
            return {
                'status': response.status_code,
                'data': response.json() if response.status_code == 200 else response.text
            }

        if response.status_code != 200:
            return {'status': response.status_code, 'data': response.text}

        aggregated_text: List[str] = []
        usage: Optional[Dict[str, Any]] = None

        print('\n🛈 Streaming response:\n')

        for line in response.iter_lines(decode_unicode=True):
            if not line:
                continue

            if line.startswith('data:'):
                payload = line[len('data:'):].strip()

                if payload == '[DONE]':
                    break

                try:
                    chunk = json.loads(payload)
                except json.JSONDecodeError:
                    continue

                if 'usage' in chunk:
                    usage = chunk['usage']

                choices = chunk.get('choices', [])
                if not choices:
                    continue

                delta = choices[0].get('delta', {}) or {}
                content = delta.get('content')
                if content:
                    print(content, end='', flush=True)
                    aggregated_text.append(content)

        print('\n')

        data = {
            'choices': [
                {
                    'message': {
                        'content': ''.join(aggregated_text)
                    }
                }
            ],
            'usage': usage,
            'streamed': True,
        }

        return {'status': 200, 'data': data}


def show_help():
    """Show help message"""
    print("""
🤖 GitHub Copilot API Client (Python)

Usage:
  python copilot_client.py <command> [options]

Commands:
  models                     - Get available models
  chat "<message>"          - Send a chat message
  chat-file <file.txt>      - Send content from a file
  
Options:
  --max-tokens <number>     - Maximum tokens (default: 1000)
  --temperature <number>    - Temperature 0-1 (default: 0.3)
  --stream                  - Enable streaming

Examples:
  python copilot_client.py models
  python copilot_client.py chat "Write a Python function to sort a list"
  python copilot_client.py chat "Explain recursion" --max-tokens 500 --temperature 0.7
  python copilot_client.py chat-file prompt.txt

Note: Make sure to set your YOUR_ACCESS_TOKEN in config.json first!
    """)


def main():
    """Main function"""
    parser = argparse.ArgumentParser(description='GitHub Copilot API Client', add_help=False)
    parser.add_argument('command', nargs='?', help='Command to execute')
    parser.add_argument('message_or_file', nargs='?', help='Message or file path')
    parser.add_argument('--max-tokens', type=int, default=1000, help='Maximum tokens')
    parser.add_argument('--temperature', type=float, default=0.3, help='Temperature')
    parser.add_argument('--stream', action='store_true', help='Enable streaming')
    parser.add_argument('--help', '-h', action='store_true', help='Show help')

    args = parser.parse_args()

    if args.help or not args.command:
        show_help()
        return

    client = CopilotAPI()

    try:
        if args.command == 'models':
            print('📋 Getting available models...')
            response = client.get_models()
            print('Models:', json.dumps(response['data'], indent=2))

        elif args.command == 'chat':
            if not args.message_or_file:
                print('❌ Please provide a message for chat')
                return

            options = {
                'max_tokens': args.max_tokens,
                'temperature': args.temperature,
                'stream': args.stream
            }

            print('💬 Sending chat message...')
            response = client.chat_completion([
                {'role': 'user', 'content': args.message_or_file}
            ], options)

            if response['status'] == 200:
                if args.stream and response['data'].get('streamed'):
                    full_text = response['data']['choices'][0]['message']['content']
                    if full_text:
                        print('\n\n📝 Final Response (aggregated):')
                        print(full_text)

                    usage = response['data'].get('usage') or {}
                    total_tokens = usage.get('total_tokens')
                    if total_tokens is not None:
                        print(f"\n📊 Usage: {total_tokens} tokens")
                else:
                    print('\n📝 Response:')
                    print(response['data']['choices'][0]['message']['content'])

                    usage = response['data'].get('usage') or {}
                    total_tokens = usage.get('total_tokens')
                    if total_tokens is not None:
                        print(f"\n📊 Usage: {total_tokens} tokens")
            else:
                print(f'❌ Error: {response["data"]}')

        elif args.command == 'chat-file':
            if not args.message_or_file:
                print('❌ Please provide a file path')
                return

            try:
                with open(args.message_or_file, 'r', encoding='utf-8') as f:
                    file_content = f.read()

                print('📁 Sending file content...')
                
                options = {
                    'max_tokens': args.max_tokens,
                    'temperature': args.temperature,
                    'stream': args.stream
                }

                response = client.chat_completion([
                    {'role': 'user', 'content': file_content}
                ], options)

                if response['status'] == 200:
                    print('\n📝 Response:')
                    print(response['data']['choices'][0]['message']['content'])
                    print(f"\n📊 Usage: {response['data']['usage']['total_tokens']} tokens")
                else:
                    print(f'❌ Error: {response["data"]}')

            except Exception as e:
                print(f'❌ Error reading file: {e}')

        else:
            print(f'❌ Unknown command: {args.command}')
            show_help()

    except Exception as e:
        print(f'❌ Error: {e}')


if __name__ == '__main__':
    main()
